### PR TITLE
fix(gnovm): proper memory allocation and garbage collection

### PR DESCRIPTION
**Pull request by ltzmaxwell**
main work in this PR:

- fixes a flaw in GC counting logic to prevent cycles that could lead to stack overflow.
- fix some allocation and GC counting logic to make sure(as much as possible) the "should not happen" won't happen, which is a issue that exists in the old GC and allocation strategy.
- The above issue isn’t fully resolved. `StringValue` is still double-counted regardless of reuse, but this will be addressed later since it’s considered an irregular case.

Thanks to @piux2 and @jefft0 to identify the initial issue. 🙏

### CODE REVIEW

**Comment by Gno2D2** [comment]
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: ltzmaxwell/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

File: gnovm/pkg/gnolang/garbage_collector.go [hunk 1]
```diff
@@ -140,17 +136,24 @@ func GCVisitorFn(gcCycle int64, alloc *Allocator, visitCount int64) Visitor {
...
 		}
 
+		alloc.Allocate(size)
+
+		// bump before visiting associated,
+		// this avoids infinite recurse.
+		if oo, isObject := v.(Object); isObject {
+			oo.SetLastGCCycle(gcCycle)
+		}
+
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/garbage_collector.go]: in a flow of : A -> B -> A visit, bump cycle for A first to avoid infinite loop.

**Comment by piux2** [comment]
LGTM. this should merge together with https://github.com/gnolang/gno/pull/4516

**Review by piux2 — APPROVED** [review]

File: gnovm/tests/files/alloc_8.gno [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+func recurse(n int) {
+	recurse(n + 1)
+}
+func main() {
+	recurse(0)
+}
+
+// Error:
+// should not happen, allocation limit exceeded while gc.
```
Comment (thehowl) [inline gnovm/tests/files/alloc_8.gno]: I don't get it, why does this happen in GC?

Comment (piux2) [reply gnovm/tests/files/alloc_8.gno]: After MaxAllocBytes is reached, the garbage collector resets Alloc, traverses the AST nodes and their values, and reallocates memory. Any object that cannot be visited during this traversal is excluded from Alloc—this is how memory is "freed."

During this process, another round of garbage collection can be triggered if no memory is freed, as demonstrated in this test case. MaxAllocBytes is reached again at [Line 145](https://github.com/gnolang/gno/pull/4302/files#diff-e21293300ec23fe16d93d7f44167d8e5e15c7d823cf430b4389357c1e267beceL145).

Just before maxBytes is reached, it is always greater than curBytes and less than curBytes+size. Without the fix in lines 147–151, the garbage collector would continue to run indefinitely until a Go runtime stack overflow occurs. The larger the value of MaxAllocBytes, the longer it takes to trigger the stack overflow—1M runs for about 20 seconds, while 3M runs for about 1 minute.

```
garbage_collector.go:

L147:   if maxBytes < curBytes+size {
                     return true
           }
L151:   alloc.Allocate(size)
```

File: gnovm/pkg/gnolang/garbage_collector.go [hunk 3]
```diff
@@ -200,8 +203,18 @@ func (fv *FuncValue) VisitAssociated(vis Visitor) (stop bool) {
...
+	switch v := fv.Parent.(type) {
+	case nil:
+		return
+	case *Block:
+		if v != nil {
+			stop = vis(v)
+		}
+	case RefValue:
+		stop = vis(v)
+	}
```
Comment (omarsy) [inline gnovm/pkg/gnolang/garbage_collector.go:231]: suggestion: maybe this to be consistent
```suggestion
	case RefValue:
		stop = vis(v)
	default:
        	panic(fmt.Sprintf("unexpected FuncValue.Parent type: %T", v))
	}
```

**Review by omarsy — APPROVED** [review]

**Review by thehowl — APPROVED** [review]

**Comment by ltzmaxwell** [comment]
working on the hidden issue.  should NOT merge this until it's fixed.

File: gnovm/pkg/gnolang/alloc.go [hunk 4]
```diff
@@ -342,92 +373,168 @@ func (alloc *Allocator) NewHeapItem(tv TypedValue) *HeapItemValue {
...
+	ss = allocPackage
 
-	for name := range pv.fBlocksMap {
-		allocFBlocksMap += int64(len(name)) + _allocName // key
-		allocFBlocksMap += _allocPointer                 // *Block
+	if !withRef {
+		return ss
 	}
 
-	return allocPackage + allocFNames + allocFBlocks + allocFBlocksMap
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/alloc.go]: these were not explicitly allocated, so not counting here. allocate them as well?

File: gnovm/pkg/gnolang/alloc.go [hunk 5]
```diff
@@ -342,34 +367,42 @@ func (alloc *Allocator) NewHeapItem(tv TypedValue) *HeapItemValue {
...
+		return 0
 	}
 
-	allocFBlocks = int64(len(pv.FBlocks)) * _allocValue
+	return allocPackage
+}
 
-	for name := range pv.fBlocksMap {
-		allocFBlocksMap += int64(len(name)) + _allocName // key
-		allocFBlocksMap += _allocPointer                 // *Block
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/alloc.go:361]: Remove this for now. as these values were not allocated (of course the take up memory), so not count here either. will improve in future.

File: gno.land/pkg/integration/testdata/many_inserts_stack_overflow.txtar [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+-- script/script.gno --
+package main
+import "gno.land/r/demo/boards"
+
+func main() {
+	n := 2000 
+	for i := 0; i < n; i++ {
+		boards.CreateReply(cross, boards.BoardID(1), boards.PostID(1), boards.PostID(1), "reply")
+	}
+}
```
Comment (ltzmaxwell) [inline gno.land/pkg/integration/testdata/many_inserts_stack_overflow.txtar:25]: the #4418 issue.

File: gnovm/tests/files/alloc_10b.gno [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	boards.CreateThread(cross, bid, "testthread", "testbody")
+	n := 5000
+	for i := 0; i < n; i++ {
+		boards.CreateReply(cross, bid, boards.PostID(1), boards.PostID(1), "reply")
+	}
+	println("ok")
+}
+
+// Error:
+// allocation limit exceeded
```
Comment (ltzmaxwell) [inline gnovm/tests/files/alloc_10b.gno:33]: the #4418 issue, with no gas limit, with alloc limit as the cap.

File: gnovm/tests/files/alloc_10a.gno [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+	boards.CreateThread(cross, bid, "testthread", "testbody")
+	n := 5000
+	for i := 0; i < n; i++ {
+		boards.CreateReply(cross, bid, boards.PostID(1), boards.PostID(1), "reply")
+	}
+	println("ok")
+}
+
+// Output:
+// ok
```
Comment (ltzmaxwell) [inline gnovm/tests/files/alloc_10a.gno:33]: #4418.

File: gnovm/tests/files/alloc_10.gno [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+import (
+	"gno.land/r/demo/boards" // not enough memory to import
+)
+
+func main() {
+	_ = boards.CreateBoard(cross, "testboard")
+}
+
+// Error:
+// main/alloc_10.gno:6:2-26: allocation limit exceeded
```
Comment (ltzmaxwell) [inline gnovm/tests/files/alloc_10.gno:14]: allocation exceeds while import a big package, GC cannot help.

File: gnovm/pkg/gnolang/values.go [hunk 10]
```diff
@@ -971,8 +972,11 @@ func (tv *TypedValue) IsTypedNil() bool {
 	if tv.V != nil {
 		return false
 	}
-	if tv.T != nil && tv.T.Kind() == PointerKind {
-		return true
+	if tv.T != nil {
+		switch tv.T.Kind() {
+		case SliceKind, FuncKind, MapKind, InterfaceKind, PointerKind, ChanKind:
+			return true
+		}
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/values.go:979]: typed nil. cc @thehowl for insight.

**Comment by ltzmaxwell** [comment]
this commit [fd7c300](https://github.com/gnolang/gno/pull/4302/commits/fd7c3005554279394919d615cc71c196ca8be804) introduces additional logic to improve the precision of allocation and GC counting. In theory, objects tracked by GC should be a subset of all allocated objects; this commit adds previously missed allocations. please also checkout the comments in files.

File: gnovm/tests/files/alloc_8.gno (hunk 2 above)
Comment (ltzmaxwell) [reply gnovm/tests/files/alloc_8.gno]: there are some further modifications after the state at this point, that I want to make it here rather than another PR to make this PR not a half-finished one. Sorry for the trouble. 🙏 
please checkout this comment: https://github.com/gnolang/gno/pull/4302#issuecomment-3244393915
cc :/ @piux2 @thehowl @omarsy

Comment (ltzmaxwell) [reply gnovm/tests/files/alloc_8.gno]: in short, the "should not happen" is should not happen, it is a corner case where the GC count exceeds the number present when GC starts, which is incorrect. The new commit: https://github.com/gnolang/gno/pull/4302/commits/fd7c3005554279394919d615cc71c196ca8be804 fix this by e.g. adding missed allocations and refining GC counting strategies, etc.

File: gno.land/pkg/integration/testdata/many_inserts_stack_overflow.txtar (hunk 6 above)
Comment (ltzmaxwell) [reply gno.land/pkg/integration/testdata/many_inserts_stack_overflow.txtar:25]: cc: @jefft0

File: gnovm/tests/files/alloc_10b.gno (hunk 7 above)
Comment (ltzmaxwell) [reply gnovm/tests/files/alloc_10b.gno:33]: cc: @jefft0

File: gnovm/pkg/gnolang/alloc.go (hunk 5 above)
Comment (piux2) [reply gnovm/pkg/gnolang/alloc.go:361]: @ltzmaxwell  The changes LGTM. 

I have two questions:

1) Do we have a test case for https://github.com/gnolang/gno/pull/4194#issuecomment-2837134378? 

if the value was not allocated but still has a shallow value during the run time, the invariant check should catch it in the GC, is that right?

I could not find such a check in the code.

The allocator is reset, but the old value (the actual allocation) was not captured to compare against the total value (remaining + GCed allocation) during GC.

https://github.com/gnolang/gno/blob/a874bc1e7d76e827e135621cc8312bf9e486f9fb/gnovm/pkg/gnolang/garbage_collector.go#L56

Is this check implemented somewhere else?

2) Is adding allocation for []FName and []FBlocks in PackageValue considered part of the future improvements?

File: gnovm/pkg/gnolang/nodes.go [hunk 11]
```diff
@@ -1333,16 +1333,23 @@ func NewPackageNode(name Name, path string, fset *FileSet) *PackageNode {
...
-		PkgName:    pn.PkgName,
-		PkgPath:    pn.PkgPath,
-		FNames:     nil,
-		FBlocks:    nil,
-		fBlocksMap: make(map[string]*Block),
+func (pn *PackageNode) NewPackage(alloc *Allocator) *PackageValue {
+	var pv *PackageValue
+	if pn.PkgName == "main" {
+		// Allocation is only for the new created main package,
+		// other packages are allocted while loading from store.
```
Comment (piux2) [inline gnovm/pkg/gnolang/nodes.go:1340]: Where is package value allocation counted when loading from the store? Can you point it out?
This is the only place where AllocPackageValue() is called, through alloc.NewPackageValue(pn).

File: gnovm/pkg/gnolang/alloc.go (hunk 5 above)
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/alloc.go:361]: thank you for the review!

1. No the check logic you mentioned is not included in this PR, it will be shipped with the 3 take in my plan;
2. yes, that it is.

File: gnovm/pkg/gnolang/nodes.go (hunk 11 above)
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/nodes.go:1340]: The main package is built for a file test, a MsgCall (see vm/keeper.go), or MsgRun. It is not stored, but it will still be scanned during GC.

File: gnovm/pkg/gnolang/alloc.go (hunk 5 above)
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/alloc.go:361]: So this PR fills the gap between allocation and GC counting as much as possible, but it hasn’t fully achieved the newBytes <= oldBytes comparison yet, and there are still some minor details to address, like the second one, and `StringValue` counting, etc.

File: gnovm/pkg/gnolang/nodes.go (hunk 11 above)
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/nodes.go:1340]: and for other packages, see loadObjectSafe().
https://github.com/gnolang/gno/blob/a874bc1e7d76e827e135621cc8312bf9e486f9fb/gnovm/pkg/gnolang/store.go#L495-L498
